### ISSUE TITLE
“FS0058: Possible incorrect indentation” on function composition after running Fantomas

### ISSUE THREAD

**Issue by lydell** [fsprojects/fantomas#1341]
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAKAZmAfDsA5gJRjoBGFlV1Ntd9DjTzLrbZANGNEZB%2BFADOiAC4iAlgDsCg1GADaAiKAiqoAEhEBPAA7I0kKSL5LVkdQDcAhgBsArvoWm1ABn5q1kAJKSkkkQDK4gBeyO4eEAAszmAAuqYAvuHKMebaenKG-iYR5tb2jooREACMyR6QALJWAB4AMlKIdYjSIgAWOcUlAEwuMfFqSaYqFZq6jpBksLA2nZ6Wtg5yRcXd5Z4BiAC24nA2sJIAgiIAor4A8pgNkmExECIATg79ieUj8%2BkTUzNzZgsFyzuYAAzOszAEdFZoIgAEKITCwB6IAAKVgeVi2okQD1%2BakezwiA1UQzU7z%2Bn0y31mYLyi0KQMiYIgkAhUNh8MRTVgAHdsdArMIfBZYPyJAdcap8YgXoM3qkxhkDFSJbSAWgVhEAKxMqCs6FwhFIgCqOj0D35gskwtF4nFOswtmEMuJctyCq%2B02p8vyS3VQIAbDqWZD9RykQBhGwCwThg6CfHQESIiUQB02J2E17DeUUpWelX-X1OYpgADsQb17MNiEq2zI2JTYDTGY8RIgJNUZOZ7sp%2BZphfpJYAHBWQ1XObH9pJG83pZnZdm3bmoMr%2Bz7B8UAJyjtmHTAibGxrZbKyNqXO9uu0bLyZ9710wElkpuIHBtkGzmbHZ7O1A2cXsAOxSJdxl7H41wfP0nzKV9Kz3A8Hi-XYZl-Etz3nF1F2vUC83A%2B81WLLo1lfHw-BEc5JAAFQeLQAHVxHaGdHTnVss1JHMcJXO83XXR8ulBWCx0OB5YDsXwABFEBscQdgQs8nhYtQ20Aq8Pk4rJjAggiNQ8EpGVfaoai8TBKLaFoTnTRAAjaRERHo%2BBGJ1SI%2Bgwy8sLUxUoCMAteKgrptQM2ofEwcQanOM0rCTB4ThqHQkUEQRbWnHVNRc1iF3YkDPI0nzIMIiISkDQKagAJUQOAHngezHIZNKlLYzsOOy7ytKLHS1BKctirKir4AAOTsLZ6weS4vAPLZZB1EoAKAsAuw0G8tjsGwJGkm4ADFERPMRsUo0DWo3AqR1fHrEXgSpltWxpNoeba5KDaA2jRKEEIAfW5cQHI6GbVPJdSWvwtqgRKbdiuE9EtHOB4GnjaqOicurVGU2b5rSf7sgOviIl6INDPBqxIeh8R4wGobsVG8bJuBn73L%2B7KlpW8Q1sQG67t2-bAcOjxuhgktIHxwmYZEC7GeZ1nIvu19HuexNsXez7GJpzLsOajHOax7niL5wyADU6RhKR4CkAg4cbZylcarKJgBni8va1RugE7XanWsTEySg3fGN02EYt4CVettXbe0oFun052anE2ARAAcVEGK4sQBKkp9oFUr9uamo9PDg6BktugC7XLqZxoYX2aAAGsYXRSvREECiAgxRBJ0G5K-2YjPUZ7XCvVzrm1G6Iq%2Bb6xBuWZuERF5Fo9r0SSQskBiksOXxazJh4qZLf9XJU2nuxvVd1b87Gur5gBpRBEB0YzTJaHxG8xa5bk3jvt5RrPMht0ZfPy7njojiydiCDaKnEs5tX6-T3upA%2BfcNYD1BnzII0gbCIGEgQQaLQ7KIDIJUBAT9ihb3SphZWHls69y-nbIEwIXx80ONJAgkhXaSHdgcII9DIp2CRJRWApEMGRSSkxSynd3491yiHEswJebFH5itbEkg%2BEWC5NIVew055SEXnGARLZ6oZUtgHMCZD5jf3thAYEWspEiwkI-OoGIyDwCsJGWAiVpAjzHo0TRikkYNX9iQ-Roi87FGBE7KR4liZWDIMggBxM2gBC0P4Wo7ihFW0yC0eAlxH4BG0Mgvx-dVDAnDlIs4qSrhuIeg8GwmBICJL0SIzGR8PDAkLlIjJDxdjC1wQk1%2BAhYjhEMIIdaiU5D-gSEAA)

#### Code

```fsharp
(f >> g) (bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb, c)
```
            

#### Result

```fsharp
(f >> g) (
    bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb,
    c
)

```
            
#### Problem description

Formatting the input results in:

```
FS0058: Possible incorrect indentation, this token is offside of context ...
```

Workaround: Call `f` and `g` normally instead of composing them together first.

I found this when upgrading from Fantomas 4.0 to 4.3 at work.

#### Extra information

- [ ] The formatted result breaks by code.
- [x] The formatted result gives compiler warnings.
- [ ] I or my company would be willing to help fix this.

#### Options

Fantomas Master at 12/26/2020 15:00:23 - 76fa31efcc568685abafe6efe770557f66147917

```fsharp
    { config with
                SemicolonAtEndOfLine = true }
```

**Comment by lydell** [comment]
Here’s another case.

This time, the issue is not only the compiler warning, but also:

- Idempotency. Running Fantomas twice causes further changes.
- Redundant parens around the `if` condition?


Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKB6dYDyAnAlgOb4B2AhgDZikAOArgC4oA6JFiDYAHmALysQI%2BAGZgSsTgApRZWXPkLFS5StVr1GzWQCUYBgAtEJAYNMBGExEQUAzokumwAJlatMYAGKxcAWzIMGRHgwWEZ6TlhRA0QwMgAjWAA3RAAaahIwYW9YsDs4EmCs338WNg5uPgcRMElxKQdHQRktFta29o6GwW1daOMMxrALAatbexHHFxJIFPAoOwDSQhtUMABtOYhQRsgAEgYATxpkNEhSBhnNwT3EyjoT9avTAAZZwagASQKjBgBlfAAXsg3oMACxPMAAXSuAF8QVsIXtDsdVmcSBd4dddrcKPdVht3mZMaZIABZMhcAAypEQlKMhAMl0JTmeEOhpjhV22jiRRwekASsAoTJ52LuDwJgycxOuv0QPnwcAosBIAEEGABRArYYTUkjAiEQYSUOxs2GY7kk-Z81GC4Uym7i-GGsAAZhlEEgvxoZGgiAAQogiogAApkXBkHwcRC4EWNBi4e5mjkWxHWlGnO1xrE4vFoSWNUEeqDe30BoPeWmwADuMegZDsX0SsHrDHwKuzpgTSca7MEnNMlqxyP5WYdYtxEpdAFZi16fX7A8GAKo0Y64euNkjN1vt6bF41jZP91M7dOj2BCztQCd5x7vABsc9Li4ruEQAGEKA2bB%2BVTZu2gBhvGvI0TUQY8IAHQQh09c9bUve001zKd3gAdmfBdy2DUl5TiGNQMycDILAaCETPEcEKvccUOdd4AA5MLLJdKz-ZV9xdQ9TV7c0uTTSjM0Q69HUnOjBgATiYv1VWEQJcD-Hw-EI7sIJ4lM%2BIom1BOo5CnXzF0zFeF152Yt9EDlBUlQ7A9iLUk8NNFASoDHXTRP0wkiWMl9EBkuSLMVIVrJdFSSLIsBYJvJyBSEmi9PvQYzGlYyviQdFsBIAAVXADgAdXwRkbKPOyoNPRytOcmLXLvAtHDMd0vKw1VcFCAoABFrHwBU5OUxNVMcPsSocq0ovOYTb1QhKi2M8kuA%2BYQMsMEgNTGX59G8Bg8vgAqXVBVlitI0rhvKtEMSqibGjMWdpopL5hHwLhsHXfxvA1LgaHfGwbD3Qjpz2-reMHfjjtG2K3Pii6n2urgACVEDgXB4E27b3l20LDuHYH0TG2j3ISjCodh%2BH4AAOToHx8NwHUPkCHwVmLMw0aGjGMygHw6AoNsKBpLxigCGMMptUHqoMxjjMJ7x4FJdnOe57w-D52M52gfRw19OSAH1q3wLb9EgRnAc0lmTuxuKavMSSoaaiMDjwakAKR3Xi1R-awoi3kjZBs6xMaFk5xmq2yBt3A7YYUnyZjKmabpgz9ZgoGjbZjn8C5-UefluSBZRIXzsmTz3kgAOg5DqWk5TxA0-8bqlZViMgJjTXtcZWPyLKj2sez73JiS-OZoANXFf1SHgJYHcI53-vUg3W-5T2zxx8HJnqnuKQ8OgSCAvdB4KEfG8dna-tMAaDqZuCRvbr3cZ9qbl64VqJAAcQ4V73sQT691H4tfub8L44vHS59Ni6JwV0e7S2TjSf0ypoAAGt-S12gRwGw6VfiRk-EKMmHF3hcT6ofAGcdDZ-yQgAsGZtBBOEhvnYmiBqxl0DAwWsRhM6IHandEg%2BU9yqgKLhcOuBo5YNshPeyU8jpGxcsQ4W7wnD43zgAaUQIgGgc0FpGC%2BCgqMeoDT8KKoIwawjmYz3PuInOpgnCixvstBUNh9Af33t-N28FtJENFPPUhEAnAW3zv8EghB2BNUIGTH4OVEBxFJAgTRgxsF2N-lRJxVoXEuldEZfOqouaEBIKvdebYVT-DSf4Og74MqwBSj8fw31CrcR0cfPRp9jpiOcYA94ro86DALhzGM5A2zJEpCqQg3CKYsNIOw-8hFIku3RjU0RlUjGd1MK6buLSS5tg0ZSSMcR4BkC-LAL63iqE0JpCMgRuDJ74OnjEk2JCElLxaa1fANh4jsAsbc-QvwDjogpAc7RRyhEnJEfyIw8AdQaN%2BIcdg5yJGDFdNfFpWoAW6n2UrXAFBhB6zGSfSKtSpn1IuY0kBLTgUECAqEpAHyKlfKgnMSEIIzg2A8F9VYkSYRAA)


#### Code

```fsharp
// Original input:
let x =
    if not (f aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa) then
        1
    else
        2

// Formatted output of the above, in for a second format:
let x =
    if (not (
            f aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
        )) then
        1
    else
        2

```
            

#### Result

```fsharp
// Original input:
let x =
    if (not (
            f aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
        )) then
        1
    else
        2

// Formatted output of the above, in for a second format:
let x =
    if (not (
        f aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
    )) then
        1
    else
        2

```
            
#### Problem description

Please describe here the Fantomas problem you encountered.
Check out our [Contribution Guidelines](https://github.com/fsprojects/fantomas/blob/master/CONTRIBUTING.md#bug-reports).

#### Options

Fantomas Master at 12/26/2020 15:00:23 - 76fa31efcc568685abafe6efe770557f66147917

Default Fantomas configuration

**Comment by nojaf** [comment]
Hello @lydell, thank you for reporting this issue. I've opened an issue in the F# compiler to understand why the function composition is giving a warning in the first place.

As for the second problem, I understand from your point of view these might seem related but for me, they are separate issues.
Could you please open a new issue for the if/then/else parenthesis case. Thanks.

**Comment by knocte** [comment]
> I've opened an issue in the F# compiler 

Hey @nojaf can you provide the link of this?

BTW I have the feeling that the bug I recently reported (https://github.com/fsprojects/fantomas/issues/1374) might be a duplicate of this one (or duplicate of https://github.com/fsprojects/fantomas/issues/1349 ) and I don't know what the expected results of my bug would be.

**Comment by nojaf** [comment]
Hello Andres

> Hey @nojaf can you provide the link of this?

![image](https://user-images.githubusercontent.com/2621499/105137725-15407d00-5af4-11eb-9176-63de07d5f8b7.png)

Already there ;)

#1374 and #1349 look like duplicates but not to this bug.

**Comment by knocte** [comment]
> #1374 and #1349 look like duplicates but not to this bug.

But it seems that fixing https://github.com/dotnet/fsharp/issues/10852 would fix #1374 and #1349 too right?

**Comment by nojaf** [comment]
No, that is a bit of a stretch I believe.
The indentation problem of this issue is purely caused because the function is composed.
As you can see in the issue at the fsharp compiler, the problem of this thread occur when `(f >> g)` is used and not when it is a simple identifier.

The problem reported in #1374, is more one of alignment of the closing parenthesis and the arguments need to be further indented.

In this issue the indentation is correct and I believe the compiler warning is questionable.
